File: 0778-swim-in-rising-water/0778-swim-in-rising-water.go
func solve(grid [][]int, dp [][][]int, visited [][]bool, n, x, y, t int) int {
	if x < 0 || y < 0 || x == n || y == n || visited[x][y] {
		return math.MaxInt32
	}
	if x == n-1 && y == n-1 {
		return int(math.Max(float64(t), float64(grid[x][y])))
	}
	if dp[x][y][t] != -1 {
		return dp[x][y][t]
	}
	visited[x][y] = true
	currMax := int(math.Max(float64(t), float64(grid[x][y])))
	dp[x][y][t] = int(math.Min(
		math.Min(
			float64(solve(grid, dp, visited, n, x-1, y, currMax)),
			float64(solve(grid, dp, visited, n, x+1, y, currMax)),
		),
		math.Min(
			float64(solve(grid, dp, visited, n, x, y+1, currMax)),
			float64(solve(grid, dp, visited, n, x, y-1, currMax)),
		),
	))
	visited[x][y] = false
	return dp[x][y][t]
}

func swimInWater(grid [][]int) int {
	n := len(grid)
	dp := make([][][]int, n)
	visited := make([][]bool, n)
	for i := range dp {
		dp[i] = make([][]int, n)
		for j := range dp[i] {
			dp[i][j] = make([]int, n*n)
			for k := range dp[i][j] {
				dp[i][j][k] = -1
			}
		}
	}
	for i := range visited {
		visited[i] = make([]bool, n)
		for j := range visited[i] {
			visited[i][j] = false
		}
	}
	return solve(grid, dp, visited, n, 0, 0, 0)
}

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
class Solution {
public:
    vector<vector<vector<int>>> dp;
    vector<vector<bool>> visited;
    int n;
    int solve(vector<vector<int>>& grid, int x, int y, int t) {
        if (x < 0 or y < 0 or x == n or y == n or visited[x][y])
            return INT_MAX;
        if (x == n - 1 and y == n - 1)
            return max(t, grid[x][y]);
        if (dp[x][y][t] != -1)
            return dp[x][y][t];
        visited[x][y] = true;
        dp[x][y][t] = min({
            solve(grid, x - 1, y, max(t, grid[x][y])),
            solve(grid, x + 1, y, max(t, grid[x][y])),
            solve(grid, x, y + 1, max(t, grid[x][y])),
            solve(grid, x, y - 1, max(t, grid[x][y])),
        });
        visited[x][y] = false;
        return dp[x][y][t];
    }
    int swimInWater(vector<vector<int>>& grid) {
        n = grid.size();
        dp.resize(n, vector<vector<int>>(n, vector<int>(n * n, -1)));
        visited.resize(n, vector<bool>(n, false));
        return solve(grid, 0, 0, 0);
    }
};